compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -463,38 +463,91 @@ fn layout_of_uncached<'tcx>(
                 ));
             }
 
-            tcx.mk_layout(
-                cx.layout_of_struct_or_enum(
+            let get_discriminant_type =
+                |min, max| Integer::repr_discr(tcx, ty, &def.repr(), min, max);
+
+            let discriminants = if def.is_enum() {
+                def.discriminants(tcx).map(|(v, d)| (v, d.val as i128)).collect()
+            } else {
+                Vec::new()
+            };
+
+            let dont_niche_optimize_enum = def.repr().inhibit_enum_layout_opt()
+                || def
+                    .variants()
+                    .iter_enumerated()
+                    .any(|(i, v)| v.discr != ty::VariantDiscr::Relative(i.as_u32()));
+
+            let maybe_unsized = def.is_struct()
+                && def.non_enum_variant().fields.raw.last().is_some_and(|last_field| {
+                    let param_env = tcx.param_env(def.did());
+                    !tcx.type_of(last_field.did).subst_identity().is_sized(tcx, param_env)
+                });
```

Comment:
This will now pass `always_sized = true` to `layout_of_struct_or_enum` for enums and structs with no fields, but it shouldn't change behavior.